README

Jaque
=====

An experimental Q-JSGI framework, based on Node, Jack, and Q
promises.  Jaque provides adapters for routing, responders,
sessions, static file trees, common content types, and
middleware.  The Q-HTTP package provides the basic server
component.

I would have called this ``q-jack`` but ``jaque`` made me
chuckle inside.


The General Idea
================

    var JAQUE = require("jaque");
    var Q = require("q"); // from npm
    var HTTP = require("q-http"); // from npm

    var app = JAQUE.Branch({
        "": JAQUE.Method({
            "GET": function (request) {
                return {
                    "status": 200,
                    "headers": {"content-type": "text/plain"},
                    "body": ["Hello, World!\r\n"]
                };
            }
        })
    }, JAQUE.notFound);

    app = JAQUE.Decorators([
        JAQUE.Log,
        JAQUE.Error,
        JAQUE.ContentLength
    ], app);

    var server = HTTP.Server(app);
    var listening = server.listen(8080);
    Q.when(listening, function () {
        console.log("Listening on 8080.");
    }, Q.error);


Installation
============

    npm install jaque q q-http


The API
=======

End(app, notFound)

    Makes a  Q-JSGI app that only responds when there is
    nothing left on the path to route.  If the there is
    unprocessed data on the path, the returned app either
    forwards to the `notFound` app or returns a `404 Not
    Found` response.

    @param {App} app a Q-JSGI application to respond to this
    end of the routing chain.

    @param {App} notFound (optional) defaults to the
    `notFound` app.

    @returns {App}

Branch(paths, notFound)

    Makes a Q-JSGI app that branches requests based on the
    next unprocessed path component.

    @param {Object * App} paths a mapping from path
    components (single file or directory names) to Q-JSGI
    applications for subsequent routing.  The mapping may be
    a plain JavaScript `Object` record, which must own the
    mapping properties, or an object that has `has(key)` and
    `get(key)` methods in its prototype chain.

    @param {App} notFound a Q-JSGI application that handles
    requests for which the next file name does not exist in
    paths.

    @returns {App}

Method(methods, methodNotAllowed_opt)

    Routes based on HTTP method.

    @param {Object * App} methods
    @param {App} notAllowed (optional)
    @returns {App}

Accept(types, notAcceptable)

    Routes based on content negotiation, between the
    request's `accept` header and the application's list of
    possible content types.

    @param {Object * App} types mapping content types to
    apps that can handle them.
    @param {App} notAcceptable
    @returns {App}

FirstFound(apps)

    Returns the response of the first application that
    returns a non-404 resposne status.

    @param {Array * App} apps a cascade of applications to try
    successively until one of them returns a non-404 status.
    @returns {App}


Sessions
--------

CookieSession(Session)

    Creates a persistent session associated with the HTTP
    client's cookie.  These sessions are intended to persist
    for the duration that a user visits your site in the
    same browser.
    
    @param {Function(session):App} Session a function that
    creates a new Q-JSGI application for each new session.
    @returns {App}
    
PathSession(Session)

    A Q-JSGI application that creates a session associated
    with a unique path.  These sessions are intended to
    persist for the duration that a user remains in a single
    browser window.

    @param {Function(session):App} a function that creates a
    new Q-JSGI application for each new session.  It
    receives an object with the session's `id` and
    `lastAccess` `Date`.
    @returns {App}


Content
-------

Content(body, contentType="text/plain", status=200)

    Makes an app that returns a response with static content
    from memory.

    @param {Body} body a Q-JSGI response body
    @param {String} contentType
    @param {Number} status
    @returns {App} a Q-JSGI app

ContentLength(app)

    Decorates a Q-JSGI application, such that responses have
    a `"content-length"` header.  This causes the response
    content to be accumulated instead of streamed.  If a
    response already has a `"content-length"` or
    `"transfer-encoding"` header, it is passed through
    unaltered.

    @param {App} app
    @returns {App}

File(path, contentType)

    Hosts a single file, at a given path, with the given
    content type.

    @param {String} path
    @param {String} contentType
    @returns {App}

FileTree(root, options)

    Hosts a file tree.

    /!\ WARNING: does not yet verify that the canonical
    path of the root contains the canonical path of the
    requested file.

    @param {String} path
    @param {{notFound, file, directory, contentType}} options
    @returns {App}

FileConcat(fileNames, conctentType)

    Provides an app that responds with a concatenated
    file from a list of file paths.
    
    @param {Array*String} paths
    @param {String} contentType default`"text/plain"`
    @returns {App}
    
PermanentRedirect(path, status_opt)

    Creates an App that unconditionally responds with
    a permanent redirect to the given path.

    @param {String} path
    @param {Number} status
    @returns {App}
    
TemporaryRedirect(path, status=307)
    
    Creates an App that unconditionally responds with
    a temporary redirect to the given path.

    @param {String} path
    @param {Number} status (optional) default is `307`
    @returns {App}


Middleware
----------

Error(app)

    Decorates a JSGI application such that rejected response
    promises get translated into `500` server error responses
    with no content.

    @param {App} app
    @returns {App}

Log(app)

    Decorates a Q-JSGI application such that all requests and
    responses are logged.

    @param {App} app
    @returns {App}

Decorators(decorators, app)

    Wraps a Q-JSGI application in a sequence of decorators.
    @param {Array * Decorator} decorators
    @param {App} app
    @returns {App}


Adapters
--------

Json(app)

    Wraps a Q-JSGI application such that the child
    application may simply return an object, which will in
    turn be serialized into a Q-JSGI response.

    @param {Function(Request):Object} app an application
    that accepts a request and returns a JSON serializable
    object.
    @returns {App}

PostContent(app)

    Wraps an app such that it expects to receive content
    in the request body and passes that content as a string
    to as the second argument to the wrapped JSGI app.

    @param {Function(Request, Object):Response} app
    @returns {App}

PostJson(app, badRequest_opt)

    Wraps an app such that it expects to receive a JSON
    object in the HTTP POST body, and passes that object to
    the child app as a second argument.

    @param {Function(Request, Object):Response} app
    @param {App} badRequest
    @returns {App}

Inspect(callback(Request))

    @param {Function(Request):Object}
    @returns {App}


Responders
----------

badRequest(request)
    {App} an application that returns a 400 response.
    
notFound(request)
    {App} an application that returns a 404 response.

methodNotAllowed(request)
    {App} an application that returns a 405 response.

notAcceptable(request)
    {App} an application that returns a 405 response.

ok(content="", contentType="text/plain", status=200)

    Returns a Q-JSGI response with the given content.

    @param {Body} content (optional) defaults to `[""]`
    @param {String} contentType (optional) defaults to
    `"text/plain"`
    @param {Number} status (optional) defaults to `200`
    @returns {Response}

file(request, path, contentType)
    
    Constructs a JSGI response for a given path,
    handling E-Tags and client-cache hits.

    @param {Request} request
    @param {String} path
    @param {String} contentType
    @returns {Response}

redirect(location, status)

    @param {String} location
    @param {Number} status (optional) default is `301`
    @returns {Response}

json(content, options)

    @param {Object} content data to serialize as JSON
    @param {{tabs}} options
    @returns {Response}



CHANGES

DEPRECATED
----------

The following exports have been renamed.  The old names will disappear
in the next major, backward-incompatible release.

 - ``End`` is now ``Cap``
 - ``PostContent`` is now ``ContentRequest``
 - ``PostJson`` is now ``JsonRequest``


History
-------

0.1.21
 - Syntax fix that should have not made it to publication without testing.

0.1.20
 - Added support for redirecting entire subtrees where
   symbolic links are involved, optionally permanently.

0.1.19
 - Added an option to FileTree: redirectSymbolicLinks.
 - Synchronized dependencies for FS.readLink.

0.1.18
 - Synchronized dependencies for bug fix in IO stream close method.

0.1.17
 - Switched internal MIME extension checker to use the "mime"
   NPM package, which appears to be more thorough on close inspection.
 - Took advantage of Q-HTTP's ability to handle a promised body
   in the static file endware.

0.1.16
 - Synchronized dependency for q-fs bug fix for binary encoding.

0.1.12
 - Added ``Headers`` middleware for adding headers unspecified by an
   application.
 - Coordinated dependency revisions.
 - Fixed a bug in the file server; 404 responses are now delivered promptly.

0.1.11
 - Added ``text/cache-manifest`` to the ``.manifest`` MIME mapping.

0.1.10
 - WARNING: Removed ``ContentLength`` until further notice.  It is
   presently broken.
 - Consolidated all content negotiation under a single implementation
   and expanded the negotiable parameters to include:
    - ``ContentType``
    - ``Language``
    - ``Charset``
    - ``Encoding``
    - ``Host``
 - Host "negotiation" is now predicated on the host name and the port,
   separated by colons, and negotiation wildcards are accepted.
 - Added ``Select`` negotiation, wherein an app is selected and
   returned by a given function.
 - Added alternate log function and timestamp function parameters
   to Log middleware.
 - Redirectors now accept a default redirect to the same path, which
   is useful for a ``GET`` following a ``PUT`` or ``POST``.
 - Aliased the ``ok`` application maker to ``content``.  Both names
   will remain going forward.
 - Applications now receive and are obligated to route the response
   object in addition to the request object.
 - Renamed End route to Cap route. End is now deprecated.
 - Renamed PostContent and PostJson to ContentRequest and JsonRequest
   respectively.  The old names are depreacted aliases.
 - Added ``Time`` Middleware
 - Added ``visitor`` and ``tabs`` parameters, as in
   ``JSON.stringify``, to the ``Json`` middleware.
 - Withdrew ambition to implement pattern matcher; someone else's job.
 - Withdrew ambition to implement FileOverlay; better supported by
   chaining FileTree middleware end to end.

0.1.8

 - Improved the logging format

0.1.7

 - replaced an assertion with a rejection, where
   Branch routers receive bad pathInfo.

0.1.6

 - added FileConcat content app

0.1.5

 - revised in coordination with q-http to be closer
   to JSGI compliance, using scriptName and pathInfo.
 - added request inspector test server.
 - fixed the README example code.
 
0.1.4

 - synced dependencies

0.1.3

 - attempted to bring all dependencies into the new
   world of n-util

0.1.2

 - Swapped util for n-util in response to new Node module
   name conflict.

0.1.0

 - Reorganized for NPM

0.0.1

 - Renamed ComponentMap back to Branch
 - Renamed Redirect to PermanentRedirect
 - Introduced TemporaryRedirect
 - Introduced PostJson


lib/jaque/route.js

/**
    Multiplexing Routing:
        Node
        Branch
        Patterns NYI
        Method
        Accept
        Language
    Trial Routing:
        FirstFound
    Decorators:
        ContentLength
        Error
        Log
        CookieSession
        PathSession
        Decorators
    Adapters:
        PostJson
        Json
        Inspect
    Producers:
        Content
        File
        FileTree
        FileOverlay NYI
        PermanentRedirect
        TemporaryRedirect
    Producer Functions:
        ok
        badRequest
        notFound
        methodNotAllowed
        notAcceptable
        file
        directory
        redirect
        json
    Utility:
        etag
        
*/


// narwhal
var Q = require("narwhal/promise-util");
var FS = require("narwhal/q-fs");
var N_UTIL = require("narwhal/util");
var MIME = require("narwhal/mime");
var UUID = require("narwhal/uuid");
var URL = require("url"); // node
var COOKIE = require("./cookie"); // jaque

// node
var SYS = require("sys");

// jaque
var J_UTIL = require("./util");
var MIME_TYPES = require("./mime-types");

exports.Node = function (app, notFound) {
    return exports.Branch({"": app}, notFound);
};

exports.Branch = function (paths, notFound) {
    if (!paths)
        paths = {};
    if (!notFound)
        notFound = exports.notFound;
    return function (request) {
        if (!request.location)
            request.location = "/";
        if (!request.course)
            request.course = request.path;
        if (!/^\//.test(request.course)) {
            // XXX TODO
        }
        var path = request.course.substring(1);
        var parts = path.split("/");
        var part = decodeURIComponent(parts.shift());
        if (N_UTIL.has(paths, part)) {
            request.location = request.location + part + "/";
            request.course = path.substring(part.length);
            return N_UTIL.get(paths, part)(request);
        }
        return notFound(request);
    };
};

exports.Patterns = function (patterns, notFound) {
    return Q.reject("pattern routing not yet implemented");
};

exports.Content = function (body, contentType, status) {
    return function (request) {
        return {
            "status": status,
            "headers": {
                "content-type": contentType
            },
            "body": body
        };
    };
};

exports.ok = function (content, contentType, status) {
    status = status || 200;
    content = content || [""];
    contentType = contentType || "text/plain";
    return {
        "status": status,
        "headers": {
            "content-type": contentType
        },
        "body": content
    };
};

exports.ContentLength = function (app) {
    return function (request) {
        return Q.when(app(request), function (response) {
            if (
                response.headers["content-length"] ||
                response.headers["transfer-encoding"]
            ) {
                return response;
            } else {
                return Q.when(response.body, function (body) {
                    if (Array.isArray(body)) {
                        return Q.when(Q.all(body), function (body) {
                            response.body = body;
                            response.headers["content-length"] = 
                                String(body.reduce(function (sum, value) {
                                    return sum + value.length;
                                }, 0));
                            return response;
                        });
                    } else if (true /* XXX */ || !body.read) {
                        return response;
                    } else {
                        // TODO BUG XXX (does not work for binary data)
                        return Q.when(body.read(), function (body) {
                            response.body = [body];
                            response.headers["content-length"] =
                                String(body.length);
                            return response;
                        });
                    }
                });
            }
        });
    };
};

exports.File = function (path, contentType) {
    return function (request) {
        return exports.file(request, path, contentType);
    };
};

exports.FileTree = function (root, options) {
    if (!options)
        options = {};
    options.notFound = options.notFound || exports.notFound;
    options.file = options.file || exports.file;
    options.directory = options.directory || exports.directory;
    return function (request) {
        if (!request.course)
            request.course = request.path;
        var path = FS.join(root, request.course.substring(1));
        // TODO FS.contains(FS.canonical(root), FS.canonical(path))
        return Q.when(FS.stat(path), function (stat) {
            if (!stat) {
                return options.notFound(request);
            } else if (stat.isFile()) {
                return options.file(request, path, options.contentType);
            } else if (stat.isDirectory()) {
                return options.directory(request, path, options.contentType);
            } else {
                return options.notFound(request);
            }
        }, function (reason) {
            return options.notFound(request);
        });
    };
};

exports.FileOverlay = function (roots, options) {
    throw new Error("NYI");
};

exports.file = function (request, path, contentType) {
    contentType = MIME_TYPES.mimeType(
        FS.extension(path),
        contentType
    );
    return Q.when(FS.stat(path), function (stat) {
        var etag = exports.etag(stat);
        if (etag === request.headers['if-none-match'])
            return J_UTIL.responseForStatus(304);
        return Q.when(FS.open(path), function (stream) {
            return {
                "status": 200,
                "headers": {
                    "content-type": contentType,
                    "etag": etag
                },
                "body": stream
            };
        });
    });
};

exports.etag = function (stat) {
    return [
        stat.ino,
        stat.size,
        Date.parse(stat.mtime)
    ].join("-");
};

exports.directory = function (request, path) {
    return Q.reject("directory listing not yet implemented");
};

exports.PermanentRedirect = function (path, status) {
    return function (request) {
        var location = URL.resolve(request.url, path);
        return exports.redirect(location, status);
    };
};

exports.TemporaryRedirect = function (path, status) {
    return function (request) {
        var location = URL.resolve(request.url, path);
        return exports.redirect(location, status || 307);
    };
};

exports.redirect = function (location, status) {
    status = status || 301;
    // TODO assure that the location is fully qualified

    return {
        "status": status,
        "headers": {
            "location": location,
            "content-type": "text/html"
        },
        "body": [
            'Go to <a href="' + location + '">' +
            location +
            "</a>"
        ]
    };
};

/// branch on HTTP method
exports.Method = function (methods, methodNotAllowed) {
    var keys = Object.keys(methods);
    if (!methodNotAllowed)
        methodNotAllowed = exports.methodNotAllowed;
    return function (request) {
        var method = request.method;
        if (N_UTIL.has(keys, method)) {
            return N_UTIL.get(methods, method)(request);
        } else {
            return methodNotAllowed(request);
        }
    };
};

/// branch on HTTP content negotiation
exports.Accept = function (types, notAcceptable) {
    var keys = Object.keys(types);
    if (!notAcceptable)
        notAcceptable = exports.notAcceptable;
    return function (request) {
        var accept = request.headers["accept"] || "";
        var type = MIME.bestMatch(keys, accept);
        if (N_UTIL.has(keys, type)) {
            return types[type](request);
        } else {
            return notAcceptable(request);
        }
    };
};

/// branch on Language
exports.Language = function (languages, noLanguage) {
    return Q.reject("Language branching not yet implemented");
};

exports.Error = function (app) {
    return function (request) {
        return Q.when(app(request), function (response) {
            return response;
        }, function (error) {
            return J_UTIL.responseForStatus(500, error);
        });
    };
};

exports.Log = function (app) {
    return function (request) {
        SYS.log(
            "=>     " + 
            request.method + " " +
            request.path + " "
        );
        return Q.when(app(request), function (response) {
            SYS.log(
                "<= " + 
                response.status + " " +
                request.method + " " +
                request.path + " " +
                (response.headers["content-length"] || "?")
            );
            return response;
        });
    };
};

// apply a series of decorators
exports.Decorators = function (decorators, app) {
    N_UTIL.reversed(decorators).forEach(function (Middleware) {
        app = Middleware(app);
    });
    return app;
};

exports.Json = function (app) {
    return function (request) {
        return Q.when(app(request), function (object) {
            return exports.json(object);
        });
    };
};

exports.json = function (content, options) {
    options = options || {};
    try {
        var json = JSON.stringify(content, null, options.tabs);
    } catch (exception) {
        return Q.reject(exception);
    }
    return exports.ok([json]);
};

exports.PostContent = function (app) {
    return function (request) {
        return Q.when(request.body, function (body) {
            return Q.when(body.read(), function (body) {
                return app(request, body);
            });
        });
    };
};

exports.PostJson = function (app, badRequest) {
    if (!badRequest)
        badRequest = exports.badRequest;
    return exports.PostContent(function (request, body) {
        try {
            var object = JSON.parse(body);
        } catch (error) {
            return badRequest(request, error);
        }
        return app(request, object);
    });
};

exports.Inspect = function (app) {
    return exports.Method({"GET": function (request) {
        return Q.when(app(request), function (object) {
            return {
                "status": 200,
                "headers": {
                    "content-type": "text/plain"
                },
                "body": [SYS.inspect(object)]
            }
        });
    }});
};

exports.CookieSession = function (Session) {
    var sessions = {};
    function nextUuid() {
        while (true) {
            var uuid = UUID.uuid();
            if (!N_UTIL.has(sessions, uuid))
                return uuid;
        }
    }
    return function (request) {
        var cookie = COOKIE.parse(request.headers["cookie"]);
        var sessionIds = cookie["session.id"];
        if (!Array.isArray(sessionIds))
            sessionIds = [sessionIds];
        sessionIds = sessionIds.filter(function (sessionId) {
            return N_UTIL.has(sessions, sessionId);
        });
        // verifying cookie
        if (/^\/~session\//.test(request.course)) {
            if (cookie["session.id"])
                return exports.TemporaryRedirect("../")(request);
            // TODO more flexible session error page
            return {
                "status": 404,
                "headers": {
                    "content-type": "text/plain"
                },
                "body": [
                    "Access requires cookies"
                ]
            }
        // session exists
        } else if (
            N_UTIL.has(cookie, "session.id") &&
            sessionIds.length
        ) {
            var session = sessions[sessionIds[0]];
            session.lastAccess = new Date();
            request.session = session;
            return session.route(request);
        // new session
        } else {
            var session = {
                "id": nextUuid(),
                "lastAccess": new Date()
            };
            sessions[session.id] = session;
            session.route = Session(session);
            var response = exports.TemporaryRedirect(request.location + "~session/")(request);
            response.headers["set-cookie"] = COOKIE.format(
                "session.id", session.id, {
                    "path": request.location
                }
            );
            return response;
        }
    };
};

exports.PathSession = function (Session) {
    var sessions = {};
    function nextUuid() {
        while (true) {
            var uuid = UUID.uuid();
            if (!N_UTIL.has(sessions, uuid))
                return uuid;
        }
    }
    return function (request) {
        // TODO request.course and request.location
        if (request.course == "/") {
            // new session
            var session = {
                "id": nextUuid(),
                "lastAccess": new Date()
            };
            sessions[session.id] = session;
            session.route = Session(session);
            return exports.Json(function (request) {
                return session;
            })(request);
        } else if (N_UTIL.has(sessions, request.course.substring(1))) {
            return N_UTIL.get(sessions, request.course.substring(1)).route(request);
        } else {
            return exports.responseForStatus(404, "Session does not exist");
        }
    };
};

exports.FirstFound = function (cascade) {
    return function (request) {
        var i = 0, ii = cascade.length;
        function next() {
            var response = cascade[i++](request);
            if (i < ii) {
                return Q.when(response, function (response) {
                    if (response.status === 404) {
                        return next();
                    } else {
                        return response;
                    }
                });
            } else {
                return response;
            }
        }
        return next();
    };
};

exports.badRequest = J_UTIL.appForStatus(400);
exports.notFound = J_UTIL.appForStatus(404);
exports.methodNotAllowed = J_UTIL.appForStatus(405);
exports.notAcceptable = J_UTIL.appForStatus(406);

